pkg/embed/append.go

```diff
@@ -0,0 +1,175 @@
+// returns nil and no error.
+func ReadEmbeddedKotsRelease(bin string) (io.Reader, error) {
+	rewind, size, err := CalculateRewind(bin)
+	if err != nil {
+		return nil, fmt.Errorf("unable to calculate rewind: %w", err)
+	} else if rewind == 0 {
+		return nil, nil
+	}
+	fp, err := os.Open(bin)
+	if err != nil {
+		return nil, fmt.Errorf("unable to open file %s: %w", bin, err)
+	}
+	defer fp.Close()
+	if _, err := fp.Seek(-int64(rewind), io.SeekEnd); err != nil {
+		return nil, fmt.Errorf("unable to seek to end of file: %w", err)
+	}
+	data := make([]byte, size)
+	if _, err := fp.Read(data); err != nil {
+		return nil, fmt.Errorf("unable to read end of file: %w", err)
+	}
+	return bytes.NewReader(data), nil
+}
+
+// ReadAndParseEmbeddedKotsRelease reads the embedded release from the given file path.
+// Parses the release into an afero.FS and returns it. If no embedded data is found,
+// returns nil as afero.FS and no error.
+func ReadAndParseEmbeddedKotsRelease(bin string) (afero.Fs, error) {
+	fp, err := ReadEmbeddedKotsRelease(bin)
+	if err != nil {
+		return nil, fmt.Errorf("unable to read embedded data: %w", err)
+	} else if fp == nil {
+		return nil, nil
+	}
+	gzreader, err := gzip.NewReader(fp)
+	if err != nil {
+		return nil, fmt.Errorf("unable to create gzip reader: %w", err)
+	}
+	defer gzreader.Close()
+	tarreader := tar.NewReader(gzreader)
+	fs := afero.NewMemMapFs()
+	for {
+		header, err := tarreader.Next()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, fmt.Errorf("unable to read tar header: %w", err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			err = fs.MkdirAll(header.Name, os.FileMode(header.Mode))
+			if err != nil {
+				return nil, fmt.Errorf("unable to create dir: %w", err)
+			}
+		case tar.TypeReg:
+			outfile, err := fs.Create(header.Name)
+			if err != nil {
+				return nil, fmt.Errorf("unable to create file: %w", err)
+			}
+			defer outfile.Close()
+			if _, err = io.Copy(outfile, tarreader); err != nil {
+				return nil, fmt.Errorf("unable to copy file: %w", err)
+			}
+		}
+	}
+	return fs, nil
+}
+
+// Build builds the binary with the embedded release. Returns an io.ReadCloser
+// from where the binary can be read. Returns the ReadCloser, the total size of
+// the binary and an error if any.
+func Build(bin string, release string) (io.ReadCloser, int64, error) {
+	binfp, err := os.Open(bin)
+	if err != nil {
+		return nil, 0, fmt.Errorf("unable to open binary file: %w", err)
+	}
+	binstat, err := binfp.Stat()
+	if err != nil {
+		return nil, 0, fmt.Errorf("unable to stat binary file: %w", err)
+	}
+	relfp, err := os.Open(release)
+	if err != nil {
+		binfp.Close()
+		return nil, 0, fmt.Errorf("unable to open release file: %w", err)
+	}
+	relstat, err := relfp.Stat()
+	if err != nil {
+		binfp.Close()
+		relfp.Close()
+		return nil, 0, fmt.Errorf("unable to stat release file: %w", err)
+	}
+	// blocks releases larger than 100MB.
+	if relstat.Size() > 104857600 {
```

Comment:
the endmark format is `%s%010d` - IMO this implies that the size should be all `9`s and not this number?